# Carpobrotus_root_microbiome

The scripts starting with "dada" were used to generate ASVs and annotate taxonomically until getting phyloseq objects for every exeprimental dataset downloaded from the ENA. The scripts called "merged_dataset" contain all the Generalized impacts across locations analysis. 
The modelling script for the random forest analysis was adapted from Qian et al., 2020 and Liu et al., 2021. Analysis and figure layout for multi-habitat integrative analysis were built upon the recommendations of Liu et al., 2021 and Li et al., 2022.  

Qian, X., Liu, Y. X., Ye, X., Zheng, W., Lv, S., Mo, M., ... & Lu, M. (2020). Gut microbiota in children with juvenile idiopathic arthritis: characteristics, biomarker identification, and usefulness in clinical prediction. BMC genomics, 21(1), 286.
Liu, Y.-X., Qin, Y., Chen, T., Lu, M., Qian, X., Guo, X., & Bai, Y. (2021). A practical guide to
amplicon and metagenomic analysis of microbiome data. Protein & Cell, 12(5), 315–
330. https://doi.org/10.1007/s13238-020-00724-8
Li, C., Bo, H., Song, B., Chen, X., Cao, Q., Yang, R., Ji, S., Wang, L., & Liu, J. (2022). Reshaping
of the soil microbiome by the expansion of invasive plants: Shifts in structure,
diversity, co-occurrence, niche breadth, and assembly processes. Plant and Soil,
477(1–2), 629–646. https://doi.org/10.1007/s11104-022-05445-x

